ksqldb-engine/src/main/java/io/confluent/ksql/statement/SourcePropertyInjector.java

```diff
@@ -0,0 +1,123 @@
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.statement;
+
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.properties.with.CreateSourceAsProperties;
+import io.confluent.ksql.parser.properties.with.CreateSourceProperties;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.CreateSource;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.util.ErrorMessageUtil;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+
+/**
+ * An injector which injects information into {@code CreateSourceProperties}
+ * or {@code CreateSourceAsProperties}.
+ *
+ * <p>The only information injected currently is that new queries using the
+ * protobuf format should use the new default converter behavior to unwrap
+ * wrapped primitives. This injector causes new queries to use the new behavior
+ * while old (existing) queries will continue to use the old behavior.
+ *
+ * <p>If a statement that is not {@code CreateAsSelect} or {@code CreateSource}
+ * is passed in, this results in a no-op that returns the incoming statement.</p>
+ */
+public class SourcePropertyInjector implements Injector {
+
+  @SuppressWarnings("unchecked")
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement
+  ) {
+    if (!(statement.getStatement() instanceof CreateSource)
+        && !(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    try {
+      if (statement.getStatement() instanceof CreateAsSelect) {
+        return (ConfiguredStatement<T>) injectForCreateAsSelect(
+            (ConfiguredStatement<? extends CreateAsSelect>) statement);
+      } else {
+        return (ConfiguredStatement<T>) injectForCreateSource(
+            (ConfiguredStatement<? extends CreateSource>) statement);
+      }
+    } catch (final KsqlStatementException e) {
+      throw e;
+    } catch (final KsqlException e) {
+      throw new KsqlStatementException(
+          ErrorMessageUtil.buildErrorMessage(e),
+          statement.getStatementText(),
+          e.getCause());
+    }
+  }
+
+  private ConfiguredStatement<? extends CreateSource> injectForCreateSource(
+      final ConfiguredStatement<? extends CreateSource> original
+  ) {
+    final CreateSource statement = original.getStatement();
+    final CreateSourceProperties properties = statement.getProperties();
+
+    final CreateSourceProperties injectedProps = properties.withUnwrapProtobufPrimitives(true);
```

Comment:
I thought we need use KsqlConfig setting to avoid making changes in QTT to distinguish old/new QTT tests? But I didn't see where you made QTT changes to distinguish those. Did you manage to make QTT work by just adding the injector?

Reply:
That's correct. Turns out the place in QTT that I was worried about being shared between new test execution and historic test execution was not shared after all. That was my bad. (Good news is I didn't have to add the config anymore. Bad news is that means there are two completely different codepaths in QTT for executing new code versus old code...)